src/foam/nanos/auth/resetPassword/ResetPasswordTokenService.js

```diff
@@ -42,129 +43,135 @@ foam.CLASS({
       name: 'generateTokenWithParameters',
-      javaCode:
-`
-AppConfig appConfig = (AppConfig) x.get("appConfig");
-String url = appConfig.getUrl().replaceAll("/$", "");
-
-// The context passed to us won't have a user in it because obviously the user
-// isn't logged in if they're resetting their password. However, decorators on
-// DAOs we access down the line from here will want to use the user from the
-// context. Therefore we put the system user in the context here so that
-// decorators down the line won't throw NPEs when trying to access the user in
-// the context.
-User systemUser = ((Subject) getX().get("subject")).getUser();
-Subject subject = new Subject.Builder(x).setUser(systemUser).build();
-x = x.put("subject", subject);
-
-DAO userDAO = (DAO) getLocalUserDAO();
-DAO tokenDAO = (DAO) getTokenDAO();
-
-// check if email invalid
-if ( user == null || ! Email.isValid(user.getEmail()) ) {
-  throw new RuntimeException("Invalid Email");
-}
-
-Sink sink = new ArraySink();
-sink = userDAO.where(MLang.EQ(User.EMAIL, user.getEmail()))
-   .limit(1).select(sink);
-
-List list = ((ArraySink) sink).getArray();
-if ( list == null || list.size() == 0 ) {
-  throw new UserNotFoundException();
-}
-
-user = (User) list.get(0);
-if ( user == null ) {
-  throw new UserNotFoundException();
-}
-
-Token token = new Token();
-token.setUserId(user.getId());
-token.setExpiry(generateExpiryDate());
-token.setData(UUID.randomUUID().toString());
-token = (Token) tokenDAO.put(token);
-
-EmailMessage message = new EmailMessage();
-message.setTo(new String[] { user.getEmail() });
-
-HashMap<String, Object> args = new HashMap<>();
-args.put("name", String.format("%s %s", user.getFirstName(), user.getLastName()));
-args.put("link", url +"?token=" + token.getData() + "#reset");
-
-EmailsUtility.sendEmailFromTemplate(x, user, message, "reset-password", args);
-
-return true;`
+      javaCode: `
+        AppConfig appConfig = (AppConfig) x.get("appConfig");
+        String url = appConfig.getUrl().replaceAll("/$", "");
+
+        // The context passed to us won't have a user in it because obviously the user
+        // isn't logged in if they're resetting their password. However, decorators on
+        // DAOs we access down the line from here will want to use the user from the
+        // context. Therefore we put the system user in the context here so that
+        // decorators down the line won't throw NPEs when trying to access the user in
+        // the context.
+        User systemUser = ((Subject) getX().get("subject")).getUser();
+        Subject subject = new Subject.Builder(x).setUser(systemUser).build();
+        x = x.put("subject", subject);
+
+        DAO userDAO = (DAO) getLocalUserDAO();
+        DAO tokenDAO = (DAO) getTokenDAO();
+
+        // check if email invalid
+        if ( user == null || ! Email.isValid(user.getEmail()) ) {
+          throw new RuntimeException("Invalid Email");
+        }
+
+        Sink sink = new ArraySink();
```

Comment:
Why not do a find()?

Reply:
user = (User) userDAO.find(MLang.EQ(User.EMAIL, user.getEmail());
replace all of the sink setup, List, and user list.get(0). 

but email is not unique, so you may get many. What is the intent here? To find one user, or send to all users with that email?

Reply:
Hmm Joel that presents an interesting question, on reset password functionality.

If the user forgot their password,
 then all we have to go on is their email. 
 As you say there are multiple users with the same email so we actually don't care who the user is, the application should handle just sending a forgot password email out based on this action.
 
  so ```EmailsUtility.sendEmailFromTemplate``` should handle passing in null as the user argument.

Also was looking into this and saw this file is used with foam resetPasswordService, nanopay uses net.nanopay.auth.NanopayResetPasswordTokenService ... they look identical in this method.

Lets leave this service clean up till https://nanopay.atlassian.net/browse/NP-5518
and then just get the emailUtitlity service handle null users here

Reply:
Also, the forgot/reset password service needs to be updated to take the username or email.

Reply:
re: multiple users - in the end it doesn't matter. We sending one email to that address. It doesn't matter which user it's associated with.

Reply:
ResetPassword  clean up will be addressed https://nanopay.atlassian.net/browse/NP-5518